File: ios/Yappy/Data/Ratchet.swift
import CryptoKit
import Foundation

/// The Double Ratchet, byte-for-byte the same as the web and Android clients'.
///
/// See apps/webapp/src/lib/ratchet.ts for what it is and why: a symmetric chain
/// that steps once per message, so a used key is destroyed and the past cannot
/// be reopened; and a DH ratchet that steps once per reply, so a compromise
/// heals instead of lasting forever.
///
/// Everything here is pure — a session goes in, a new session comes out,
/// nothing touches disk or the network. That is deliberate: this is the part
/// where a mistake is silent and permanent. Storage is in `E2EStore`, the
/// envelope in `Cipher`.
enum Ratchet {
    private static let domain = "yappy.ratchet.v2"

    /// Beyond this many missing messages in one chain, assume something is wrong.
    private static let maxSkip = 1000
    /// And beyond this many stored keys, stop remembering — the map is unbounded otherwise.
    private static let maxSkippedKept = 2000

    // ── shapes ───────────────────────────────────────────────────────────────

    /// A session, as it is stored: base64 and numbers, nothing else, so it
    /// survives a round trip through a file without a custom serialiser.
    struct Session: Codable {
        /// The device at the other end. One session per device, never per person.
        var deviceId: String
        var rootKey: String
        var myRatchetPrivate: String
        var myRatchetPublic: String
        /// Nil until the first message from them arrives.
        var theirRatchet: String?
        var sendChain: String?
        var recvChain: String?
        var sendCount: Int = 0
        var recvCount: Int = 0
        var previousSendCount: Int = 0
        /// Keys for messages that have not arrived yet, by `theirRatchet|n`.
        ///
        /// A message that overtakes another must still open, and the key that
        /// opens it only exists on the way past. Kept here, deleted the moment
        /// it is used — this is the one place a ratchet holds a usable key for
        /// longer than an instant, and it is bounded for that reason.
        var skipped: [String: String] = [:]
        /// The X3DH preamble, repeated on every message until they answer.
        var pending: PreKeyHeader?

        enum CodingKeys: String, CodingKey {
            case deviceId, rootKey, myRatchetPrivate, myRatchetPublic, theirRatchet
            case sendChain, recvChain, sendCount, recvCount, previousSendCount, skipped, pending
        }

        init(
            deviceId: String,
            rootKey: String,
            myRatchetPrivate: String,
            myRatchetPublic: String,
            theirRatchet: String? = nil,
            sendChain: String? = nil,
            recvChain: String? = nil,
            sendCount: Int = 0,
            recvCount: Int = 0,
            previousSendCount: Int = 0,
            skipped: [String: String] = [:],
            pending: PreKeyHeader? = nil
        ) {
            self.deviceId = deviceId
            self.rootKey = rootKey
            self.myRatchetPrivate = myRatchetPrivate
            self.myRatchetPublic = myRatchetPublic
            self.theirRatchet = theirRatchet
            self.sendChain = sendChain
            self.recvChain = recvChain
            self.sendCount = sendCount
            self.recvCount = recvCount
            self.previousSendCount = previousSendCount
            self.skipped = skipped
            self.pending = pending
        }

        init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            deviceId = c.get(.deviceId, "")
            rootKey = c.get(.rootKey, "")
            myRatchetPrivate = c.get(.myRatchetPrivate, "")
            myRatchetPublic = c.get(.myRatchetPublic, "")
            theirRatchet = c.opt(.theirRatchet)
            sendChain = c.opt(.sendChain)
            recvChain = c.opt(.recvChain)
            sendCount = c.get(.sendCount, 0)
            recvCount = c.get(.recvCount, 0)
            previousSendCount = c.get(.previousSendCount, 0)
            skipped = c.get(.skipped, [:])
            pending = c.opt(.pending)
        }
    }

    /// What the first messages carry so the other end can derive the same root.
    struct PreKeyHeader: Codable {
        /// The sender's ephemeral public key, this session only.
        var ek: String
        /// Which of the recipient's prekeys it was agreed against.
        var spkId: Int
        var otkId: Int?

        enum CodingKeys: String, CodingKey { case ek, spkId, otkId }

        init(ek: String, spkId: Int, otkId: Int?) {
            self.ek = ek
            self.spkId = spkId
            self.otkId = otkId
        }

        init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            ek = c.get(.ek, "")
            spkId = c.get(.spkId, 1)
            otkId = c.opt(.otkId)
        }

        /// `otkId` is written even when it is nil: the other two clients read a
        /// missing key as a parse failure rather than as "no one-time prekey".
        func encode(to encoder: Encoder) throws {
            var out = encoder.container(keyedBy: CodingKeys.self)
            try out.encode(ek, forKey: .ek)
            try out.encode(spkId, forKey: .spkId)
            try out.encode(otkId, forKey: .otkId)
        }
    }

    /// What every message carries so the ratchet can be followed.
    struct Header {
        /// The sender's current ratchet public key.
        var dh: String
        /// How long the previous sending chain was, so gaps in it can be closed.
        var pn: Int
        /// Position in the current chain.
        var n: Int
        var pre: PreKeyHeader?
    }

    /// This device's private halves, as `DeviceKeys` holds them.
    struct Privates {
        var deviceId: String
        var userId: String
        var identityPrivate: String
        var signedPreKeyId: Int
        var signedPreKeyPrivate: String
        var preKeys: [Int: String]
    }

    // ── key derivation ───────────────────────────────────────────────────────

    /// The root step: mix a fresh Diffie-Hellman into the root key and get a new
    /// root key and a new chain key out.
    ///
    /// The old root key is the salt rather than the input, which is what makes
    /// this a ratchet rather than a chain of hashes.
    private static func rootStep(_ rootKey: Data, _ dh: Data) -> (Data, Data) {
        let out = hkdf(ikm: dh, salt: rootKey, info: Data("\(domain).root".utf8), count: 64)
        return (out.prefix(32), out.suffix(32))
    }

    /// The chain step: one message key, and the chain key that replaces this one.
    ///
    /// Two different constants so the message key can never be walked forward
    /// into the chain.
    private static func chainStep(_ chainKey: Data) -> (Data, Data) {
        let key = SymmetricKey(data: chainKey)
        let messageKey = Data(HMAC<SHA256>.authenticationCode(for: Data([1]), using: key))
        let nextChain = Data(HMAC<SHA256>.authenticationCode(for: Data([2]), using: key))
        return (messageKey, nextChain)
    }

    /// A message key is a key and a nonce; neither is ever reused, so neither travels.
    private static func messageCipher(_ messageKey: Data) -> (SymmetricKey, Data) {
        // No salt, which RFC 5869 defines as HashLen zero bytes — the other two
        // clients pass nothing and get exactly this.
        let out = hkdf(
            ikm: messageKey,
            salt: Data(repeating: 0, count: 32),
            info: Data("\(domain).message".utf8),
            count: 44
        )
        return (SymmetricKey(data: out.prefix(32)), out.suffix(12))
    }

    /// The X3DH agreement, from either side. Two DHs, or one when no prekey was free.
    private static func initialRoot(_ dhSpk: Data, _ dhOtk: Data?) -> Data {
        let ikm = dhOtk.map { dhSpk + $0 } ?? dhSpk
        let salt = Data(SHA256.hash(data: Data("\(domain).salt".utf8)))
        return hkdf(ikm: ikm, salt: salt, info: Data("\(domain).x3dh".utf8), count: 32)
    }

    private static func hkdf(ikm: Data, salt: Data, info: Data, count: Int) -> Data {
        HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: ikm),
            salt: salt,
            info: info,
            outputByteCount: count
        ).withUnsafeBytes { Data($0) }
    }

    private static func agree(_ privateKey: Data, _ publicKey: Data) throws -> Data {
        let mine = try Curve25519.KeyAgreement.PrivateKey(rawRepresentation: privateKey)
        let theirs = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: publicKey)
        return try mine.sharedSecretFromKeyAgreement(with: theirs).withUnsafeBytes { Data($0) }
    }

    // ── starting a session ───────────────────────────────────────────────────

    /// Begin talking to a device that has never talked back.
    ///
    /// The recipient's signed prekey stands in as its first ratchet key. That is
    /// the whole trick behind an encrypted message to somebody whose phone is
    /// off: their half of the first DH is something they published in advance
    /// and still hold.
    static func initiate(bundle: KeyBundle) throws -> Session {
        guard let spk = Data(base64Encoded: bundle.signedPreKey.key) else {
            throw Cipher.CipherError.malformed
        }

        let ephemeral = Curve25519.KeyAgreement.PrivateKey()
        let root = try initialRoot(
            agree(ephemeral.rawRepresentation, spk),
            bundle.oneTimePreKey
                .flatMap { Data(base64Encoded: $0.key) }
                .map { try agree(ephemeral.rawRepresentation, $0) }
        )

        // The first DH ratchet step happens immediately, so the very first
        // message is already under a key neither the prekey nor the ephemeral
        // can reproduce.
        let myRatchet = Curve25519.KeyAgreement.PrivateKey()
        let (rootKey, sendChain) = try rootStep(root, agree(myRatchet.rawRepresentation, spk))

        return Session(
            deviceId: bundle.deviceId,
            rootKey: rootKey.base64EncodedString(),
            myRatchetPrivate: myRatchet.rawRepresentation.base64EncodedString(),
            myRatchetPublic: myRatchet.publicKey.rawRepresentation.base64EncodedString(),
            theirRatchet: bundle.signedPreKey.key,
            sendChain: sendChain.base64EncodedString(),
            pending: PreKeyHeader(
                ek: ephemeral.publicKey.rawRepresentation.base64EncodedString(),
                spkId: bundle.signedPreKey.id,
                otkId: bundle.oneTimePreKey?.id
            )
        )
    }

    /// Build the other side of a session from the preamble on an incoming
    /// message.
    ///
    /// Nil when this device does not hold the prekeys the sender used: a pool
    /// that has been rotated, or a message for somebody else.
    static func accept(_ pre: PreKeyHeader, me: Privates, theirDeviceId: String) -> Session? {
        guard pre.spkId == me.signedPreKeyId else { return nil }
        let otkPrivate = pre.otkId.flatMap { me.preKeys[$0] }
        if pre.otkId != nil, otkPrivate == nil { return nil }

        do {
            guard let ephemeral = Data(base64Encoded: pre.ek),
                  let spkPrivate = Data(base64Encoded: me.signedPreKeyPrivate)
            else { return nil }

            let root = try initialRoot(
                agree(spkPrivate, ephemeral),
                otkPrivate.flatMap { Data(base64Encoded: $0) }.map { try agree($0, ephemeral) }
            )

            // The signed prekey is this end's first ratchet key, matching what
            // the sender assumed. The first DH step happens when their message
            // is read.
            let key = try Curve25519.KeyAgreement.PrivateKey(rawRepresentation: spkPrivate)
            return Session(
                deviceId: theirDeviceId,
                rootKey: root.base64EncodedString(),
                myRatchetPrivate: me.signedPreKeyPrivate,
                myRatchetPublic: key.publicKey.rawRepresentation.base64EncodedString()
            )
        } catch {
            return nil
        }
    }

    // ── sending ──────────────────────────────────────────────────────────────

    struct Sealed {
        var session: Session
        var header: Header
        var ciphertext: String
    }

    /// One message, and the session that replaces this one.
    ///
    /// The session is returned rather than mutated so a failed send cannot
    /// advance the ratchet: a chain that has stepped past a message nobody
    /// received is a conversation that never recovers.
    static func encrypt(_ session: Session, _ plaintext: String, aad: Data) -> Sealed? {
        guard let chain = session.sendChain, let chainData = Data(base64Encoded: chain) else {
            return nil
        }

        let (messageKey, nextChain) = chainStep(chainData)
        let header = Header(
            dh: session.myRatchetPublic,
            pn: session.previousSendCount,
            n: session.sendCount,
            pre: session.pending
        )

        do {
            let (key, nonce) = messageCipher(messageKey)
            let box = try ChaChaPoly.seal(
                Data(plaintext.utf8),
                using: key,
                nonce: ChaChaPoly.Nonce(data: nonce),
                authenticating: aad
            )
            var next = session
            next.sendChain = nextChain.base64EncodedString()
            next.sendCount += 1
            return Sealed(
                session: next,
                header: header,
                ciphertext: (box.ciphertext + box.tag).base64EncodedString()
            )
        } catch {
            return nil
        }
    }

    // ── receiving ────────────────────────────────────────────────────────────

    /// Walk a chain forward, keeping the keys for messages that have not arrived.
    ///
    /// Messages overtake each other — a phone that was asleep, two sockets, a
    /// retry. Every key stepped over is one that opens a message still in
    /// flight, so it is kept until it is used. The cap guards against a header
    /// claiming a message number far in the future.
    private static func skipTo(_ session: Session, until: Int) -> Session? {
        guard let chain = session.recvChain, let theirs = session.theirRatchet,
              var chainData = Data(base64Encoded: chain)
        else { return session }
        if until - session.recvCount > maxSkip { return nil }

        var next = session
        var n = session.recvCount
        while n < until {
            let (messageKey, nextChain) = chainStep(chainData)
            next.skipped["\(theirs)|\(n)"] = messageKey.base64EncodedString()
            chainData = nextChain
            n += 1
        }

        // Oldest first, so what falls off the end is least likely to arrive.
        let overflow = next.skipped.count - maxSkippedKept
        if overflow > 0 {
            for key in next.skipped.keys.sorted().prefix(overflow) { next.skipped[key] = nil }
        }

        next.recvChain = chainData.base64EncodedString()
        next.recvCount = n
        return next
    }

    /// A DH ratchet step: their new key arrived, so both chains start again.
    private static func turn(_ session: Session, _ theirRatchet: String) throws -> Session {
        guard let theirKey = Data(base64Encoded: theirRatchet),
              let rootKeyData = Data(base64Encoded: session.rootKey),
              let myPrivate = Data(base64Encoded: session.myRatchetPrivate)
        else { throw Cipher.CipherError.malformed }

        let (rootAfterReceive, recvChain) = try rootStep(rootKeyData, agree(myPrivate, theirKey))

        // A new keypair of our own, so the reply travels under a secret they
        // cannot derive from anything they have seen. This is the step that
        // heals.
        let myRatchet = Curve25519.KeyAgreement.PrivateKey()
        let (rootKey, sendChain) = try rootStep(
            rootAfterReceive,
            agree(myRatchet.rawRepresentation, theirKey)
        )

        var next = session
        next.previousSendCount = session.sendCount
        next.sendCount = 0
        next.recvCount = 0
        next.theirRatchet = theirRatchet
        next.rootKey = rootKey.base64EncodedString()
        next.myRatchetPrivate = myRatchet.rawRepresentation.base64EncodedString()
        next.myRatchetPublic = myRatchet.publicKey.rawRepresentation.base64EncodedString()
        next.recvChain = recvChain.base64EncodedString()
        next.sendChain = sendChain.base64EncodedString()
        return next
    }

    struct Opened {
        var session: Session
        var plaintext: String
    }

    /// The message, and the session that replaces this one — or nil.
    ///
    /// Nil is every failure: a header from a chain this session cannot reach, a
    /// key that was already used, a tag that does not check. None of them are
    /// distinguished, and none of them advance the ratchet.
    static func decrypt(_ session: Session, _ header: Header, _ ciphertext: String, aad: Data) -> Opened? {
        guard let body = Data(base64Encoded: ciphertext), body.count > 16 else { return nil }

        func open(_ messageKey: Data) -> String? {
            do {
                let (key, nonce) = messageCipher(messageKey)
                let box = try ChaChaPoly.SealedBox(
                    nonce: ChaChaPoly.Nonce(data: nonce),
                    ciphertext: body.prefix(body.count - 16),
                    tag: body.suffix(16)
                )
                return String(data: try ChaChaPoly.open(box, using: key, authenticating: aad), encoding: .utf8)
            } catch {
                return nil
            }
        }

        // A message that arrived late, whose key was kept on the way past.
        let skippedKey = "\(header.dh)|\(header.n)"
        if let stored = session.skipped[skippedKey], let key = Data(base64Encoded: stored) {
            guard let plaintext = open(key) else { return nil }
            var next = session
            next.skipped[skippedKey] = nil
            return Opened(session: next, plaintext: plaintext)
        }

        do {
            var next = session

            // Their ratchet moved on. Close out the chain we were reading
            // first: the messages we never saw from it may still turn up.
            if header.dh != session.theirRatchet {
                guard let closed = skipTo(next, until: header.pn) else { return nil }
                next = try turn(closed, header.dh)
            }

            guard let caughtUp = skipTo(next, until: header.n),
                  let chain = caughtUp.recvChain,
                  let chainData = Data(base64Encoded: chain)
            else { return nil }

            let (messageKey, nextChain) = chainStep(chainData)
            guard let plaintext = open(messageKey) else { return nil }

            var after = caughtUp
            after.recvChain = nextChain.base64EncodedString()
            after.recvCount = caughtUp.recvCount + 1
            // They have answered, so the X3DH preamble has done its job.
            after.pending = nil
            return Opened(session: after, plaintext: plaintext)
        } catch {
            return nil
        }
    }
}

File: android/app/src/main/java/gg/yappy/app/data/Ratchet.kt
package gg.yappy.app.data

import kotlinx.serialization.Serializable
import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.generators.HKDFBytesGenerator
import org.bouncycastle.crypto.macs.HMac
import org.bouncycastle.crypto.params.HKDFParameters
import org.bouncycastle.crypto.params.KeyParameter
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import org.bouncycastle.crypto.params.X25519PublicKeyParameters
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64

/**
 * The Double Ratchet, byte-for-byte the same as the web client's.
 *
 * See apps/webapp/src/lib/ratchet.ts for what it is and why: a symmetric chain
 * that steps once per message, so a used key is destroyed and the past cannot
 * be reopened; and a DH ratchet that steps once per reply, so a compromise
 * heals instead of lasting forever.
 *
 * Everything here is pure — a session goes in, a new session comes out, nothing
 * touches disk or the network. That is deliberate: this is the part where a
 * mistake is silent and permanent, so it is the part that has to be testable
 * off a device. Storage is in [E2EStore], the envelope in [Cipher].
 */
object Ratchet {

    private const val DOMAIN = "yappy.ratchet.v2"

    /** Beyond this many missing messages in one chain, assume something is wrong. */
    private const val MAX_SKIP = 1000

    /** And beyond this many stored keys, stop remembering — the map is unbounded otherwise. */
    private const val MAX_SKIPPED_KEPT = 2000

    private val random = SecureRandom()

    // ── shapes ───────────────────────────────────────────────────────────────

    /**
     * A session, as it is stored: base64 and numbers, nothing else, so it
     * survives a round trip through a file without a custom serialiser.
     */
    @Serializable
    data class Session(
        /** The device at the other end. One session per device, never per person. */
        val deviceId: String,
        val rootKey: String,
        val myRatchetPrivate: String,
        val myRatchetPublic: String,
        /** Null until the first message from them arrives. */
        val theirRatchet: String? = null,
        val sendChain: String? = null,
        val recvChain: String? = null,
        val sendCount: Int = 0,
        val recvCount: Int = 0,
        val previousSendCount: Int = 0,
        /**
         * Keys for messages that have not arrived yet, by `${theirRatchet}|${n}`.
         *
         * A message that overtakes another must still open, and the key that
         * opens it only exists on the way past. Kept here, deleted the moment it
         * is used — this map is the one place a ratchet holds a usable key for
         * longer than an instant, and it is bounded for that reason.
         */
        val skipped: Map<String, String> = emptyMap(),
        /**
         * The X3DH preamble, repeated on every message until they answer.
         *
         * A conversation begins with one side talking to a prekey bundle. Until
         * a reply proves the other end built the same session, every message has
         * to carry enough to build it — otherwise losing the first message
         * strands the conversation permanently.
         */
        val pending: PreKeyHeader? = null,
    )

    /** What the first messages carry so the other end can derive the same root. */
    @Serializable
    data class PreKeyHeader(
        /** The sender's ephemeral public key, this session only. */
        val ek: String,
        /** Which of the recipient's prekeys it was agreed against. */
        val spkId: Int,
        val otkId: Int? = null,
    )

    /** What every message carries so the ratchet can be followed. */
    data class Header(
        /** The sender's current ratchet public key. */
        val dh: String,
        /** How long the previous sending chain was, so gaps in it can be closed. */
        val pn: Int,
        /** Position in the current chain. */
        val n: Int,
        val pre: PreKeyHeader?,
    )

    // ── key derivation ───────────────────────────────────────────────────────

    /**
     * The root step: mix a fresh Diffie-Hellman into the root key and get a new
     * root key and a new chain key out.
     *
     * The old root key is the salt rather than the input, which is what makes
     * this a ratchet rather than a chain of hashes: an attacker who learns the
     * DH output still needs the root, and one who learns the root still needs
     * the DH.
     */
    private fun rootStep(rootKey: ByteArray, dh: ByteArray): Pair<ByteArray, ByteArray> {
        val out = hkdf(dh, rootKey, "$DOMAIN.root".toByteArray(), 64)
        return out.copyOfRange(0, 32) to out.copyOfRange(32, 64)
    }

    /**
     * The chain step: one message key, and the chain key that replaces this one.
     *
     * Two different constants so the message key can never be walked forward
     * into the chain — knowing a message key must say nothing about the next.
     */
    private fun chainStep(chainKey: ByteArray): Pair<ByteArray, ByteArray> =
        hmac(chainKey, byteArrayOf(1)) to hmac(chainKey, byteArrayOf(2))

    /** A message key is a key and a nonce; neither is ever reused, so neither travels. */
    private fun messageCipher(messageKey: ByteArray): Pair<ByteArray, ByteArray> {
        val out = hkdf(messageKey, null, "$DOMAIN.message".toByteArray(), 44)
        return out.copyOfRange(0, 32) to out.copyOfRange(32, 44)
    }

    /** The X3DH agreement, from either side. Two DHs, or one when no prekey was free. */
    private fun initialRoot(dhSpk: ByteArray, dhOtk: ByteArray?): ByteArray {
        val ikm = if (dhOtk == null) dhSpk else dhSpk + dhOtk
        val salt = MessageDigest.getInstance("SHA-256").digest("$DOMAIN.salt".toByteArray())
        return hkdf(ikm, salt, "$DOMAIN.x3dh".toByteArray(), 32)
    }

    // ── starting a session ───────────────────────────────────────────────────

    /**
     * Begin talking to a device that has never talked back.
     *
     * The recipient's signed prekey stands in as its first ratchet key. That is
     * the whole trick behind an encrypted message to somebody whose phone is
     * off: their half of the first DH is something they published in advance
     * and still hold.
     */
    fun initiate(bundle: KeyBundle): Session {
        val ephemeral = X25519PrivateKeyParameters(random)
        val spk = unb64(bundle.signedPreKey.key)

        val root = initialRoot(
            agree(ephemeral.encoded, spk),
            bundle.oneTimePreKey?.let { agree(ephemeral.encoded, unb64(it.key)) },
        )

        // The first DH ratchet step happens immediately, so the very first
        // message is already under a key neither the prekey nor the ephemeral
        // can reproduce.
        val myRatchet = X25519PrivateKeyParameters(random)
        val (rootKey, sendChain) = rootStep(root, agree(myRatchet.encoded, spk))

        return Session(
            deviceId = bundle.deviceId,
            rootKey = b64(rootKey),
            myRatchetPrivate = b64(myRatchet.encoded),
            myRatchetPublic = b64(myRatchet.generatePublicKey().encoded),
            theirRatchet = bundle.signedPreKey.key,
            sendChain = b64(sendChain),
            pending = PreKeyHeader(
                ek = b64(ephemeral.generatePublicKey().encoded),
                spkId = bundle.signedPreKey.id,
                otkId = bundle.oneTimePreKey?.id,
            ),
        )
    }

    /** The receiving device's private halves, as [DeviceKeys] holds them. */
    data class Privates(
        val deviceId: String,
        val userId: String,
        val identityPrivate: String,
        val signedPreKeyId: Int,
        val signedPreKeyPrivate: String,
        val preKeys: Map<Int, String>,
    )

    /**
     * Build the other side of a session from the preamble on an incoming message.
     *
     * Null when this device does not hold the prekeys the sender used: a pool
     * that has been rotated, or a message for somebody else. The caller shows
     * the unreadable state, which is the honest answer.
     */
    fun accept(pre: PreKeyHeader, me: Privates, theirDeviceId: String): Session? {
        if (pre.spkId != me.signedPreKeyId) return null
        val otkPrivate = pre.otkId?.let { me.preKeys[it] }
        if (pre.otkId != null && otkPrivate == null) return null

        return try {
            val ephemeral = unb64(pre.ek)
            val root = initialRoot(
                agree(unb64(me.signedPreKeyPrivate), ephemeral),
                otkPrivate?.let { agree(unb64(it), ephemeral) },
            )

            // The signed prekey is this end's first ratchet key, matching what
            // the sender assumed. The first DH step happens when their message
            // is read.
            val spkPrivate = X25519PrivateKeyParameters(unb64(me.signedPreKeyPrivate), 0)
            Session(
                deviceId = theirDeviceId,
                rootKey = b64(root),
                myRatchetPrivate = me.signedPreKeyPrivate,
                myRatchetPublic = b64(spkPrivate.generatePublicKey().encoded),
            )
        } catch (_: Exception) {
            null
        }
    }

    // ── sending ──────────────────────────────────────────────────────────────

    /** One message, and the session that replaces this one. */
    data class Sealed(val session: Session, val header: Header, val ciphertext: String)

    /**
     * The session is returned rather than mutated so a failed send cannot
     * advance the ratchet: a chain that has stepped past a message nobody
     * received is a conversation that never recovers.
     */
    fun encrypt(session: Session, plaintext: String, aad: ByteArray): Sealed? {
        val chain = session.sendChain ?: return null
        val (messageKey, nextChain) = chainStep(unb64(chain))
        val header = Header(
            dh = session.myRatchetPublic,
            pn = session.previousSendCount,
            n = session.sendCount,
            pre = session.pending,
        )
        val (key, nonce) = messageCipher(messageKey)
        val ciphertext = b64(Cipher.aead(key, nonce, aad, plaintext.toByteArray(), encrypt = true))

        return Sealed(
            session = session.copy(sendChain = b64(nextChain), sendCount = session.sendCount + 1),
            header = header,
            ciphertext = ciphertext,
        )
    }

    // ── receiving ────────────────────────────────────────────────────────────

    /**
     * Walk a chain forward, keeping the keys for messages that have not arrived.
     *
     * Messages overtake each other — a phone that was asleep, two sockets, a
     * retry. Every key stepped over is one that opens a message still in flight,
     * so it is kept until it is used. The cap guards against a header claiming a
     * message number far in the future, which would otherwise be an invitation
     * to derive keys until the process dies.
     */
    private fun skipTo(session: Session, until: Int): Session? {
        val chainKey = session.recvChain ?: return session
        val theirs = session.theirRatchet ?: return session
        if (until - session.recvCount > MAX_SKIP) return null

        val skipped = session.skipped.toMutableMap()
        var chain = unb64(chainKey)
        var n = session.recvCount

        while (n < until) {
            val (messageKey, nextChain) = chainStep(chain)
            skipped["$theirs|$n"] = b64(messageKey)
            chain = nextChain
            n += 1
        }

        // Oldest first, so what falls off the end is least likely to arrive.
        val overflow = skipped.size - MAX_SKIPPED_KEPT
        if (overflow > 0) skipped.keys.take(overflow).forEach { skipped.remove(it) }

        return session.copy(skipped = skipped, recvChain = b64(chain), recvCount = n)
    }

    /** A DH ratchet step: their new key arrived, so both chains start again. */
    private fun turn(session: Session, theirRatchet: String): Session {
        val theirKey = unb64(theirRatchet)
        val (rootAfterReceive, recvChain) =
            rootStep(unb64(session.rootKey), agree(unb64(session.myRatchetPrivate), theirKey))

        // A new keypair of our own, so the reply travels under a secret they
        // cannot derive from anything they have seen. This is the step that
        // heals.
        val myRatchet = X25519PrivateKeyParameters(random)
        val (rootKey, sendChain) = rootStep(rootAfterReceive, agree(myRatchet.encoded, theirKey))

        return session.copy(
            previousSendCount = session.sendCount,
            sendCount = 0,
            recvCount = 0,
            theirRatchet = theirRatchet,
            rootKey = b64(rootKey),
            myRatchetPrivate = b64(myRatchet.encoded),
            myRatchetPublic = b64(myRatchet.generatePublicKey().encoded),
            recvChain = b64(recvChain),
            sendChain = b64(sendChain),
        )
    }

    data class Opened(val session: Session, val plaintext: String)

    /**
     * The message, and the session that replaces this one — or null.
     *
     * Null is every failure: a header from a chain this session cannot reach, a
     * key that was already used, a tag that does not check. None of them are
     * distinguished, and none of them advance the ratchet, so a message that
     * cannot be opened costs nothing but itself.
     */
    fun decrypt(session: Session, header: Header, ciphertext: String, aad: ByteArray): Opened? {
        fun open(messageKey: ByteArray): String? = try {
            val (key, nonce) = messageCipher(messageKey)
            String(Cipher.aead(key, nonce, aad, unb64(ciphertext), encrypt = false))
        } catch (_: Exception) {
            null
        }

        // A message that arrived late, whose key was kept on the way past.
        val skippedKey = "${header.dh}|${header.n}"
        session.skipped[skippedKey]?.let { stored ->
            val plaintext = open(unb64(stored)) ?: return null
            return Opened(session.copy(skipped = session.skipped - skippedKey), plaintext)
        }

        return try {
            var next = session

            // Their ratchet moved on. Close out the chain we were reading
            // first: the messages we never saw from it may still turn up.
            if (header.dh != session.theirRatchet) {
                val closed = skipTo(next, header.pn) ?: return null
                next = turn(closed, header.dh)
            }

            val caughtUp = skipTo(next, header.n) ?: return null
            val chain = caughtUp.recvChain ?: return null

            val (messageKey, nextChain) = chainStep(unb64(chain))
            val plaintext = open(messageKey) ?: return null

            Opened(
                caughtUp.copy(
                    recvChain = b64(nextChain),
                    recvCount = caughtUp.recvCount + 1,
                    // They have answered, so the X3DH preamble has done its job.
                    pending = null,
                ),
                plaintext,
            )
        } catch (_: Exception) {
            null
        }
    }

    // ── primitives ───────────────────────────────────────────────────────────

    private fun hkdf(ikm: ByteArray, salt: ByteArray?, info: ByteArray, length: Int): ByteArray {
        val out = ByteArray(length)
        HKDFBytesGenerator(SHA256Digest()).apply {
            init(HKDFParameters(ikm, salt, info))
            generateBytes(out, 0, length)
        }
        return out
    }

    private fun hmac(key: ByteArray, message: ByteArray): ByteArray {
        val mac = HMac(SHA256Digest())
        mac.init(KeyParameter(key))
        mac.update(message, 0, message.size)
        return ByteArray(mac.macSize).also { mac.doFinal(it, 0) }
    }

    private fun agree(privateKey: ByteArray, publicKey: ByteArray): ByteArray {
        val out = ByteArray(32)
        X25519PrivateKeyParameters(privateKey, 0)
            .generateSecret(X25519PublicKeyParameters(publicKey, 0), out, 0)
        return out
    }

    private fun b64(bytes: ByteArray): String = Base64.getEncoder().encodeToString(bytes)
    private fun unb64(text: String): ByteArray = Base64.getDecoder().decode(text)
}
